lib/ansible/inventory/group.py

```diff
@@ -113,12 +113,21 @@ def add_child_group(self, group):
 
             self.clear_hosts_cache()
 
-    def _check_children_depth(self):
-
+    def _check_children_depth(self, exclude=None):
+        '''
+        optimization:
+        if 'exclude' set is provided, these groups will not be recursed into
+        '''
+
+        if exclude is None:
+            exclude = set([])
         try:
             for group in self.child_groups:
+                if group.name in exclude:
+                    continue
```

Comment:
This isn't right.

Sure, it's true that the exact method for the exact group has already been called, but the result of this method is can differ given different approaches. We can still `continue` here, but only in the more limited situation where the current `self.depth` is more 1 or more less than the `group.depth`. We can skip that case because if the top group hasn't increased depth by this path, that group's ancestors surely won't either. But that's more narrow than what's coded here. What's coded here is wrong.